### ISSUE TITLE
Implement sender blacklisting

### ISSUE THREAD

**Issue by jvehent** [mozilla-services/autopush#227]
We may want to block access to specific senders that violate limits or are known bad actors. This can be either implemented in the autopush server directly, or as a component of the infrastructure (we can use iptables since autopush servers have public IPs).

**Comment by bbangert** [comment]
This will be handled by ops at the ELB layer per existing guidelines.

**Comment by jvehent** [comment]
While we do want this capability, ops is currently unable to blacklist at the ELB layer. Sender blacklisting will also require calculating sending rates, which is something that would better live at the application layer.

Which guidelines are you referring to?

**Comment by bbangert** [comment]
I meant that ops already has a procedure in place to blacklist abusive IP's (which @oremj says can be done in ELB). If a request gets all the way to our server, then blacklisting them at that layer is a bit late.

If there's something specific of benefit we can change to the autopush code, then it would make sense to re-open this ticket.

**Comment by oremj** [comment]
I spoke too soon on this issue. Security groups can only whitelist traffic, but can't blacklist. We do have the ability to blacklist CIDRs via VPC network ACLs, but that will block the IP from accessing any resources within the VPC.

Accounting is also an issue, IP blocking helps in cases of service affecting levels of abuse, but there is no way to tell if sender A is legitimately sending X req/s or not.

**Comment by bbangert** [comment]
This will likely need a bloom filter that the endpoint nodes can lookup.

**Comment by jrconlin** [comment]
Closed (Overcome by Events; Abuse mostly not self-identifying and mostly due to folks not reading the return result)

**Comment by jvehent** [comment]
@jrconlin : could you summarize the status of blacklisting for both VAPID and non-VAPID senders?

**Comment by jrconlin** [comment]
From what we've seen, blacklisting is probably not going to be terribly effective. The problem is that we don't have much to go on to identify a given source if they do not use VAPID to self identify, and the original incentive model we had has been dropped due to outside constraints. In addition, the "abuse" we're seeing currently is due to publishers effectively ignoring the return result. (e.g. the current high rate of 410 and 404s.) Instead, what we're probably going to do is implement a general throttle response for publishers that do not self identify with VAPID. We may then implement a second system to whitelist good providers. 

We can, of course, throttle known bad IP blocks, but that's going to be done at an operations level.  

I'm obviously not a fan of discussing details publicly here, feel free to email me or contact me via IRC if you want more.
